src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreaker.java

```diff
@@ -0,0 +1,230 @@
+            state.counter().onFailure();
+            final EventCount count = state.counter().getCount();
+            if (checkIfExceedingFailureThreshold(count)) {
+                // changes to OPEN if failure rate exceeds the threshold
+                if (current.compareAndSet(state, newOpenState())) {
+                    logStateTransition(CircuitState.OPEN, count);
+                }
+            }
+        } else if (state.isHalfOpen()) {
+            // returns to OPEN if a request fails during HALF_OPEN
+            if (current.compareAndSet(state, newOpenState())) {
+                logStateTransition(CircuitState.OPEN, EventCount.ZERO);
+            }
+        }
+    }
+
+    private boolean checkIfExceedingFailureThreshold(EventCount count) {
+        return config.minimumRequestThreshold() <= count.total() &&
+               config.failureRateThreshold() < count.failureRate();
+    }
+
+    /**
+     * Decides whether a request should be allowed or refused according to the current circuit state.
+     */
+    boolean canRequest() {
+        final State state = current.get();
+        if (state.isClosed()) {
+            // all requests are allowed during CLOSED
+            return true;
+        } else if (state.isHalfOpen() || state.isOpen()) {
+            if (state.checkTimeout() && current.compareAndSet(state, newHalfOpenState())) {
+                // changes to HALF_OPEN if OPEN state has timed out
+                logStateTransition(CircuitState.HALF_OPEN, EventCount.ZERO);
+                return true;
+            }
+            // all other requests are refused
+            return false;
+        }
+        return true;
+    }
+
+    private State newOpenState() {
+        return new State(CircuitState.OPEN, config.circuitOpenWindow(), NoOpCounter.INSTANCE);
+    }
+
+    private State newHalfOpenState() {
+        return new State(CircuitState.HALF_OPEN, config.trialRequestInterval(), NoOpCounter.INSTANCE);
+    }
+
+    private State newClosedState() {
+        return new State(CircuitState.CLOSED, Duration.ZERO, new SlidingWindowCounter(config, clock));
+    }
+
+    private void logStateTransition(CircuitState circuitState, EventCount count) {
+        if (logger.isInfoEnabled()) {
+            final int capacity = name.length() + circuitState.name().length() + 32;
+            final StringBuilder builder = new StringBuilder(capacity);
+            builder.append("name:");
+            builder.append(name);
+            builder.append(" state:");
+            builder.append(circuitState.name());
+            if (EventCount.ZERO.equals(count)) {
+                builder.append(" fail:- total:-");
+            } else {
+                builder.append(" fail:");
+                builder.append(count.failure());
+                builder.append(" total:");
+                builder.append(count.total());
+            }
+            logger.info(builder.toString());
+        }
+    }
+
+    // visible for testing
```

Comment:
Could we use `@VisibleForTesting` annotation? If there's no dependency available, please feel free to add one.